# INICIALIZACION - EDA

In [1]:
# Importamos las primeras librerias que vamos a utilizar, gzip para leer el archivo json zipeado, json y pandas.

import gzip
import pandas as pd
import json

In [2]:
# Realizamos funcion parse que nos permite paresar la ruta del archivo json en formato .gz 
# y pedimos que nos devuelva un data frame

def parse(path):
    """
    Parse a gzip-compressed JSON file and return a DataFrame.

    Parameters:
    path (str): The path to the gzip-compressed JSON file.

    Returns:
    pd.DataFrame: A DataFrame containing the JSON data.
    """
    data = []
    with gzip.open(path, 'rt', encoding='utf-8') as file:
        for line in file:
            json_data = json.loads(line)
            data.append(json_data)
    
    df = pd.DataFrame(data)
    return df

# Carga de Datos
Vamos a comenzar con archivo de steam_games a generar nuestro Data Frame a partir de la funcion parse y vamos a realizar una primera vista de como esta compuesto el DF, que tan grande es y los tipos de datos que posee.

In [3]:
# Llama a la función parse y le pasamos el archivo Steam Games
data_frame = parse("./Base_Data/steam_games.json.gz")
# Generamos nuestro data frame para trabajar

In [4]:
# Hacemos un shape para ver el tamanio del df
data_frame.shape

(120445, 13)

In [5]:
# Al ver que es un df grande de mas de 120000 filas y 13 columnas, vamos a pedir que nos traigas las primeras 5 filas
data_frame.head() 

publisher genres app_name title  url release_date tags reviews_url specs  \
0       NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   NaN   
1       NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   NaN   
2       NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   NaN   
3       NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   NaN   
4       NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   NaN   

  price early_access   id developer  
0   NaN          NaN  NaN       NaN  
1   NaN          NaN  NaN       NaN  
2   NaN          NaN  NaN       NaN  
3   NaN          NaN  NaN       NaN  
4   NaN          NaN  NaN       NaN

In [6]:
# Al ver que tenemos todos valores nulos al inicio, vamos a pedir un .info para ver la totalidad de nulos que existen en df
data_frame.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120445 entries, 0 to 120444
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher     24083 non-null  object
 1   genres        28852 non-null  object
 2   app_name      32133 non-null  object
 3   title         30085 non-null  object
 4   url           32135 non-null  object
 5   release_date  30068 non-null  object
 6   tags          31972 non-null  object
 7   reviews_url   32133 non-null  object
 8   specs         31465 non-null  object
 9   price         30758 non-null  object
 10  early_access  32135 non-null  object
 11  id            32133 non-null  object
 12  developer     28836 non-null  object
dtypes: object(13)
memory usage: 11.9+ MB


# Limpieza y Preparacion de los Datos
En la primera vista, nos encontramos con un Data Set grande, que tiene mas de 120 mil filas y 13 columnas, tiene una gran cantidad de datos nulos, todas las variables son de tipo objeto y ocupa una buena cantidad de memoria ya que utiliza un 11.9+ MB.
A continuación vamos a seguir preparando los datos para nuestro análisis. 

In [7]:
# Eliminamos todos los datos nulos de las filas
data_games = data_frame.dropna()

In [8]:
# Realizamos un .shape para ver la cantidad de filas que nos quedaron con datos 
data_games.shape

(22530, 13)

In [9]:
# Tambien hacemos un .info para ver si tenemos algunos otros datos nulos y la memoria que estamos usando
data_games.info()

<class 'pandas.core.frame.DataFrame'>
Index: 22530 entries, 88310 to 120443
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher     22530 non-null  object
 1   genres        22530 non-null  object
 2   app_name      22530 non-null  object
 3   title         22530 non-null  object
 4   url           22530 non-null  object
 5   release_date  22530 non-null  object
 6   tags          22530 non-null  object
 7   reviews_url   22530 non-null  object
 8   specs         22530 non-null  object
 9   price         22530 non-null  object
 10  early_access  22530 non-null  object
 11  id            22530 non-null  object
 12  developer     22530 non-null  object
dtypes: object(13)
memory usage: 2.4+ MB


In [10]:
data_games

publisher  \
88310             Kotoshiro   
88311      Making Fun, Inc.   
88312          Poolians.com   
88313                  彼岸领域   
88315   Trickjump Games Ltd   
...                     ...   
120439      Bidoniera Games   
120440      Ghost_RUS Games   
120441               Sacada   
120442         Laush Studio   
120443             SIXNAILS   

                                                   genres  \
88310       [Action, Casual, Indie, Simulation, Strategy]   
88311                [Free to Play, Indie, RPG, Strategy]   
88312   [Casual, Free to Play, Indie, Simulation, Sports]   
88313                         [Action, Adventure, Casual]   
88315                     [Action, Adventure, Simulation]   
...                                                   ...   
120439                 [Action, Adventure, Casual, Indie]   
120440              [Casual, Indie, Simulation, Strategy]   
120441                          [Casual, Indie, Strategy]   
120442                        [Indie, Racing, Simulation]   
120443                                    [Casual, Indie]   

                        app_name                     title  \
88310        Lost Summoner Kitty       Lost Summoner Kitty   
88311                  Ironbound                 Ironbound   
88312    Real Pool 3D - Poolians   Real Pool 3D - Poolians   
88313                    弹炸人2222                   弹炸人2222   
88315      Battle Royale Trainer     Battle Royale Trainer   
...                          ...                       ...   
120439              Kebab it Up!              Kebab it Up!   
120440            Colony On Mars            Colony On Mars   
120441  LOGistICAL: South Africa  LOGistICAL: South Africa   
120442             Russian Roads             Russian Roads   
120443       EXIT 2 - Directions       EXIT 2 - Directions   

                                                      url release_date  \
88310   http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
88311   http://store.steampowered.com/app/643980/Ironb...   2018-01-04   
88312   http://store.steampowered.com/app/670290/Real_...   2017-07-24   
88313      http://store.steampowered.com/app/767400/2222/   2017-12-07   
88315   http://store.steampowered.com/app/772540/Battl...   2018-01-04   
...                                                   ...          ...   
120439  http://store.steampowered.com/app/745400/Kebab...   2018-01-04   
120440  http://store.steampowered.com/app/773640/Colon...   2018-01-04   
120441  http://store.steampowered.com/app/733530/LOGis...   2018-01-04   
120442  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
120443  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   

                                                     tags  \
88310       [Strategy, Action, Indie, Casual, Simulation]   
88311   [Free to Play, Strategy, Indie, RPG, Card Game...   
88312   [Free to Play, Simulation, Sports, Casual, Ind...   
88313                         [Action, Adventure, Casual]   
88315   [Action, Adventure, Simulation, FPS, Shooter, ...   
...                                                   ...   
120439        [Action, Indie, Casual, Violent, Adventure]   
120440              [Strategy, Indie, Casual, Simulation]   
120441                          [Strategy, Indie, Casual]   
120442                        [Indie, Simulation, Racing]   
120443  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...   

                                              reviews_url  \
88310   http://steamcommunity.com/app/761140/reviews/?...   
88311   http://steamcommunity.com/app/643980/reviews/?...   
88312   http://steamcommunity.com/app/670290/reviews/?...   
88313   http://steamcommunity.com/app/767400/reviews/?...   
88315   http://steamcommunity.com/app/772540/reviews/?...   
...                                                   ...   
120439  http://steamcommunity.com/app/745400/reviews/?...   
120440  http://steamcommunity.com/app/773640/reviews/?...

In [12]:
# Exportamos el data frame a un csv para luego utilizar el csv en el ETL
data_games.to_csv('steam_games.csv', index=False)

# Reviews
Realizamos pasos similares al archivo de steam_games para obtener el data frame y los datos de reviews.

In [14]:
# Este archivo los vamos a abrir de una forma diferente, ya que al normalizar los datos, no podiamos visualizarlos de buena manera si aplicabamos la funcion parse.
# Vamos a crear una lista vacia para almacenar los diccionarios y un contador para inicializar el total de lineas

data_list = []
lineas_con_errores = []
# Contadores para el total de líneas y registros leídos correctamente
total_lineas = 0
registros_correctos = 0

data_reviews = []

# Abrir el archivo y leerlo línea por línea
with open('./Base_Data/user_reviews.json', 'r', encoding='utf-8') as archivo:
    for num_linea, linea in enumerate(archivo, start=1):
        total_lineas += 1
        try:
            # Intentar cargar la línea como JSON
            data = json.loads(linea)
            data_reviews.append(data)
            registros_correctos += 1
        except json.JSONDecodeError:
            try:
                # Si falla como JSON, intentar cargarlo como diccionario
                data = eval(linea)  # Usa eval para interpretar la línea como un diccionario
                if isinstance(data, dict):
                    data_reviews.append(data)
                    registros_correctos += 1
                else:
                    lineas_con_errores.append(num_linea)
            except Exception as e:
                lineas_con_errores.append(num_linea)

# Crear un DataFrame a partir de la lista de diccionarios
df_reviews = pd.DataFrame(data_reviews)

In [15]:
# Vamos a ver que datos nos da el data frame creado con un .head
df_reviews.head()

user_id                                           user_url  \
0  76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1            js41637               http://steamcommunity.com/id/js41637   
2          evcentric             http://steamcommunity.com/id/evcentric   
3              doctr                 http://steamcommunity.com/id/doctr   
4          maplemage             http://steamcommunity.com/id/maplemage   

                                             reviews  
0  [{'funny': '', 'posted': 'Posted November 5, 2...  
1  [{'funny': '', 'posted': 'Posted June 24, 2014...  
2  [{'funny': '', 'posted': 'Posted February 3.',...  
3  [{'funny': '', 'posted': 'Posted October 14, 2...  
4  [{'funny': '3 people found this review funny',...

In [16]:
# Se transforma a columnas cada elemento de las listas y creamos nuevo data frame
df_reviews2 = pd.json_normalize(data_reviews, record_path=['reviews'])

In [17]:
# Se agrega el 'user_id', 'user_url' a las columnas separadas y concatenamos con el data frame anterior
df_reviews2 = pd.concat([df_reviews[['user_id', 'user_url']], df_reviews2], axis=1)

In [18]:
# Vemos ahora como nos quedan todos los datos separados por columnas
df_reviews2.head()

user_id                                           user_url funny  \
0  76561197970982479  http://steamcommunity.com/profiles/76561197970...         
1            js41637               http://steamcommunity.com/id/js41637         
2          evcentric             http://steamcommunity.com/id/evcentric         
3              doctr                 http://steamcommunity.com/id/doctr         
4          maplemage             http://steamcommunity.com/id/maplemage         

                      posted last_edited item_id  \
0   Posted November 5, 2011.                1250   
1      Posted July 15, 2011.               22200   
2     Posted April 21, 2011.               43110   
3      Posted June 24, 2014.              251610   
4  Posted September 8, 2013.              227300   

                                           helpful  recommend  \
0                                   No ratings yet       True   
1                                   No ratings yet       True   
2                                   No ratings yet       True   
3  15 of 20 people (75%) found this review helpful       True   
4     0 of 1 people (0%) found this review helpful       True   

                                              review  
0  Simple yet with great replayability. In my opi...  
1               It's unique and worth a playthrough.  
2  Great atmosphere. The gunplay can be a bit chu...  
3  I know what you think when you see this title ...  
4  For a simple (it's actually not all that simpl...

In [19]:
# Pedimos un .info para ver la informacion y el tamaño que tenemos en nuestro data set
df_reviews2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59305 entries, 0 to 59304
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   user_id      25799 non-null  object
 1   user_url     25799 non-null  object
 2   funny        59305 non-null  object
 3   posted       59305 non-null  object
 4   last_edited  59305 non-null  object
 5   item_id      59305 non-null  object
 6   helpful      59305 non-null  object
 7   recommend    59305 non-null  bool  
 8   review       59305 non-null  object
dtypes: bool(1), object(8)
memory usage: 3.7+ MB


In [20]:
# Se eliminan las filas con valores nulos
df_reviews2 = df_reviews2.dropna()

In [21]:
# Volemos a pedir info del data set
df_reviews2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 25799 entries, 0 to 25798
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   user_id      25799 non-null  object
 1   user_url     25799 non-null  object
 2   funny        25799 non-null  object
 3   posted       25799 non-null  object
 4   last_edited  25799 non-null  object
 5   item_id      25799 non-null  object
 6   helpful      25799 non-null  object
 7   recommend    25799 non-null  bool  
 8   review       25799 non-null  object
dtypes: bool(1), object(8)
memory usage: 1.8+ MB


In [22]:
# Ahora con un dataframe mas liviano y sin datos nulos, exportamos a un csv para luego utilizarlo en ETL
df_reviews2.to_csv('user_reviews.csv', index=False, encoding='utf-8')

# Items
Aplicamos los pasos similares para obtener los datos y data frame de Items

In [24]:
# Este archivo los vamos a abrir de una forma diferente, ya que al normalizar los datos, no podiamos visualizarlos de buena manera si aplicabamos la funcion parse.
# Vamos a crear una lista vacia para almacenar los diccionarios y un contador para inicializar el total de lineas

data_list = []
lineas_con_errores = []
# Contadores para el total de líneas y registros leídos correctamente
total_lineas = 0
registros_correctos = 0

data_items = []

# Abrir el archivo y leerlo línea por línea
with open('./Base_Data/users_items.json', 'r', encoding='utf-8') as archivo:
    for num_linea, linea in enumerate(archivo, start=1):
        total_lineas += 1
        try:
            # Intentar cargar la línea como JSON
            data = json.loads(linea)
            data_items.append(data)
            registros_correctos += 1
        except json.JSONDecodeError:
            try:
                # Si falla como JSON, intentar cargarlo como diccionario
                data = eval(linea)  # Usa eval para interpretar la línea como un diccionario
                if isinstance(data, dict):
                    data_items.append(data)
                    registros_correctos += 1
                else:
                    lineas_con_errores.append(num_linea)
            except Exception as e:
                lineas_con_errores.append(num_linea)

# Crear un DataFrame a partir de la lista de diccionarios
df_items = pd.DataFrame(data_items)

In [25]:
# Vamos a ver que datos nos da el data frame creado con un .head
df_items.head()

user_id  items_count           steam_id  \
0  76561197970982479          277  76561197970982479   
1            js41637          888  76561198035864385   
2          evcentric          137  76561198007712555   
3         Riot-Punch          328  76561197963445855   
4              doctr          541  76561198002099482   

                                            user_url  \
0  http://steamcommunity.com/profiles/76561197970...   
1               http://steamcommunity.com/id/js41637   
2             http://steamcommunity.com/id/evcentric   
3            http://steamcommunity.com/id/Riot-Punch   
4                 http://steamcommunity.com/id/doctr   

                                               items  
0  [{'item_id': '10', 'item_name': 'Counter-Strik...  
1  [{'item_id': '10', 'item_name': 'Counter-Strik...  
2  [{'item_id': '1200', 'item_name': 'Red Orchest...  
3  [{'item_id': '10', 'item_name': 'Counter-Strik...  
4  [{'item_id': '300', 'item_name': 'Day of Defea...

In [26]:
# Se transforma a columnas cada elemento de las listas y creamos nuevo data frame
df_items2 = pd.json_normalize(data_items, record_path=['items'])

In [27]:
# Se agrega el 'user_id', 'items_count', 'steam_id', 'user_url' a las columnas separadas y concatenamos con el data frame anterior
df_items2 = pd.concat([df_items[['user_id', 'items_count', 'steam_id', 'user_url']], df_items2], axis=1)

In [28]:
# Vemos ahora como nos quedan todos los datos separados por columnas
df_items2.head()

user_id  items_count           steam_id  \
0  76561197970982479        277.0  76561197970982479   
1            js41637        888.0  76561198035864385   
2          evcentric        137.0  76561198007712555   
3         Riot-Punch        328.0  76561197963445855   
4              doctr        541.0  76561198002099482   

                                            user_url item_id  \
0  http://steamcommunity.com/profiles/76561197970...      10   
1               http://steamcommunity.com/id/js41637      20   
2             http://steamcommunity.com/id/evcentric      30   
3            http://steamcommunity.com/id/Riot-Punch      40   
4                 http://steamcommunity.com/id/doctr      50   

                   item_name  playtime_forever  playtime_2weeks  
0             Counter-Strike                 6                0  
1      Team Fortress Classic                 0                0  
2              Day of Defeat                 7                0  
3         Deathmatch Classic                 0                0  
4  Half-Life: Opposing Force                 0                0

In [29]:
# Pedimos un .info para ver la informacion y el tamaño que tenemos en nuestro data set
df_items2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5153209 entries, 0 to 5153208
Data columns (total 8 columns):
 #   Column            Dtype  
---  ------            -----  
 0   user_id           object 
 1   items_count       float64
 2   steam_id          object 
 3   user_url          object 
 4   item_id           object 
 5   item_name         object 
 6   playtime_forever  int64  
 7   playtime_2weeks   int64  
dtypes: float64(1), int64(2), object(5)
memory usage: 314.5+ MB


In [30]:
# Se eliminan las filas con valores nulos
df_items2 = df_items2.dropna()

In [31]:
# Volemos a pedir info del data set
df_items2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 88310 entries, 0 to 88309
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   user_id           88310 non-null  object 
 1   items_count       88310 non-null  float64
 2   steam_id          88310 non-null  object 
 3   user_url          88310 non-null  object 
 4   item_id           88310 non-null  object 
 5   item_name         88310 non-null  object 
 6   playtime_forever  88310 non-null  int64  
 7   playtime_2weeks   88310 non-null  int64  
dtypes: float64(1), int64(2), object(5)
memory usage: 6.1+ MB


In [32]:
# Ahora con un dataframe mas liviano y sin datos nulos, exportamos a un csv para luego utilizarlo en ETL
df_items2.to_csv('users_items.csv', index=False)